libbfddp: BFD protocol handling

rzalamena opened the pull request:
Summary
------------

Implement a BFD protocol adapter library extra part. The BFD data plane library will now handle some level of protocol logic to keep the software boiler plate code level low between vendor implementations. The new extra part is not mandatory, however recommended to vendors so they avoid repeating the BFD protocol logic and every one can share the same state machine/protocol handling.

The objective is to reduce maintenance work and code duplication.

rzalamena committed:
libbfddp: import some BFD protocol definitions

We'll use some of the BFD protocol definitions in the future callback
based library implementation, so lets start the change.

While here add the missing SHUTDOWN session flag to disable a BFD
session.

rzalamena committed:
soft_bfddpd: initialize debug variables

When the session has a invalid address the debug function will print
garbage from stack.

rzalamena committed:
libbfddp: implement BFD protocol handling

Allow users to use the boiler plate BFD protocol handling code
(optionally) by including the `bfddp_extra.h` header.

The new library section implements the BFD session handling and calls
the appropriated callbacks for the hardware implementers to fill in with
their logic. This way we can have a BFD implementation that can be used
in any hardware and the only cost is implementing the hardware calls
from the library callbacks.

rzalamena committed:
soft_bfddpd: organize header inclusion

Since now we have a bunch of headers required by the library, lets
remove some of the cruft from the data plane implementation example.

rzalamena committed:
soft_bfddpd: fix two events API bugs

- Check the FD in the correct array position
- Decrement the descriptors amount on removal

rzalamena committed:
soft_bfddpd: refactory timers API

While torture testing the software BFD data plane some crashes were
found related with the timers caused by activation/deactivation on the
same `poll` run.

The timers API should now handle this situation however it had to be
reworked to support it and now it is no longer possible to receive
timers/fd actions with the callbacks return.

Extra: some debug/asserts were kept to make sure these failures never
happen again.

rzalamena committed:
libbfddp: update exported header

Fix some spelling problems and add default port.

rzalamena committed:
soft_bfddpd: use new default port

rzalamena committed:
soft_bfddpd: fix ipv6 command line parsing

rzalamena committed:
libbfddp: fix unix socket connection

UNIX sockets do not support no delay option because they don't use nagle
algorithm at all.

rzalamena committed:
soft_bfddpd: fix crash when no sessions

The lookup by packet function broke when it started returning a variable
that is not used in the RBT_FOREACH macro, so lets handle this new
situation properly.

rzalamena committed:
libbfddp: transmission interval jitter optional

Add option to generate transmission interval without jitter, otherwise
if one requires a transmission interval without jitter it would be
impossible to remove it (unless reimplementing the function which is not
desirable).

rzalamena committed:
libbfddp: export BFD control packet generator

Export a function to allow users to generate their own BFD control
packet without the requirement of using the packet send function.

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+		bfddp_callbacks.bc_state_change =
+			bfddp_session_state_change_dummy;
+
+#define CALLBACK_CHECK(cb)                                                     \
```
Define the MACRO outside of the body of the function?

rzalamena replied:
It is only used here (and it is specialized)... Any reason to "export" it?

msune replied:
I am not saying it's exported. I would keep it in the same translation unit, but having the definition of a MACRO function in the middle of the body affects readability.

rzalamena replied:
Sure, I'll move it out of the function.

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+	CALLBACK_CHECK(bc_rx_control_update);
+	CALLBACK_CHECK(bc_rx_control_stop);
+
+#undef CALLBACK_CHECK
```
Remove `#undef`
Together with the previous comment. Define a prefix (e.g. `BFDDP_CALLBACK_CHECK`). `undef` is unnecessary as `.c` will not be included (hopefully) in other translation units.

rzalamena replied:
Makes sense, saves us one line of code.

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+	 * Load flags.
+	 *
+	 * NOTE: normalize boolean values (e.g. `!!`) so packet build functions
+	 * can use it with shift (e.g. (multihop << X)).
```
Uncommon, but fine.

No need to use shift, btw; `(flags & SESSION_SHUTDOWN) > 0` would do it

rzalamena replied:
The shift part is on the packet bulding function, this is just a comment.

To avoid using `if` conditionals and generating more code lines I opted for this so we could use the `bool` for shifting like this: `bool << bit-place`.

msune replied:
My comment was about the "uncommon" (at least to me) ` !!(flags & SESSION_ECHO);`

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+	bs->bs_echo = !!(flags & SESSION_ECHO);
+
+	if (bs->bs_multihop)
+		port = htons(BFD_MULTI_HOP_PORT);
```
(they are symmetric, so it doesn't matter but..) here conversion is Host -> Network, and 20 lines before it was Network->Host

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+	/* Handle administrative shutdown. */
+	if (bs->bs_admin_shutdown) {
+		/* Send a control packet to tell we are shutting down. */
+		bs->bs_state = STATE_ADMINDOWN;
```
prefix `STATE_ADMINDOWN`? S/t like `BFDD_` or alike

msune replied:
idem for other enums / macros

rzalamena replied:
Sorry, I did not understand your comment... You are suggesting to add prefix to all enumerations already defined?

msune replied:
yup

rzalamena replied:
I think this is a bit too late for this change... It would cause tons of code change.

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+	case STATE_UP:
+		/* NOTHING: we haven't and the peer hasn't sent INIT yet. */
+		break;
+	}
```
Consider adding a `default:` with `assert(0)` and immediate exit, to catch memory errors

rzalamena replied:
This is a library, I'm not sure it'd be wise to add an assert here. If we were to add an `assert` I think it would be best to add it on "entrance" functions instead in the middle of the way, there is no way to properly return error codes here.

msune replied:
`assert()`s are perfectly fine in a library, when used properly - read here, a bug in the code. They shall _never_ be used when validating input from outside the program (read, input from a user, network, file etc..), as these errors are not bugs for the library.

In this case, as I understand it, if a value is read outside of the `case` list, then surely there is a bug; either memory corruption or an unhandled, possibly new, enum.

As you know `assert()` will not crash when not compiled with debug.

I would suggest to add a `default:` that calls `assert()`  (or better, wraps assert to log the critical condition when not compiled with `DEBUG`), and return from the function with an error code, or at least with a consistent state, so that when 
 execution can be attempted to be continued when compiled without `DEBUG`.

rzalamena replied:
> assert()s are perfectly fine in a library, when used properly - read here, a bug in the code. They shall never be used when validating input from outside the program (read, input from a user, network, file etc..), as these errors are not bugs for the library.

I agree.

> In this case, as I understand it, if a value is read outside of the case list, then surely there is a bug; either memory corruption or an unhandled, possibly new, enum.

This is handled with the warning: `-Wswitch-enum` (see `CMakeLists.txt`)

> I would suggest to add a default: that calls assert() (or better, wraps assert to log the critical condition when not compiled with DEBUG), and return from the function with an error code, or at least with a consistent state, so that when
execution can be attempted to be continued when compiled without DEBUG.

Makes sense, I'll try to fix this.

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+
+		/* Notify state change. */
+		bfddp_send_session_state_change(bs);
+		break;
```
idem

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+	case STATE_UP:
+		/* NOTHING. */
+		break;
+	}
```
idem

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+		break;
+	case STATE_UP:
+		bfddp_session_sm_up(bs, arg, nstate);
+		break;
```
idem

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+	}
+
+	/* Call application callback to notify about state change. */
+	if (ostate != bs->bs_state) {
```
Consider immediate return, to have less indentation:

```
	if (ostate == bs->bs_state) 
              return;

```

rzalamena replied:
Makes sense, I'll fix it.

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+	bs->bs_rid = ntohl(bcp->local_id);
+
+	/* Detect timers change: */
+	if (ntohl(bcp->desired_tx) != bs->bs_rtx)
```
Consider alternative:

```
timers_changed |= (ntohl(bcp->desired_tx) != bs->bs_rtx);
timers_changed |= (ntohl(bcp->required_rx) != bs->bs_rrx);
...

```

rzalamena replied:
Thanks! It should save us a few branchings (if the compiler is not doing it already). I'll fix it.

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+		return BPV_ZERO_LOCAL_ID;
+
+	/* Invalid remote ID with established session. */
+	state = (bcp->state_bits >> 6);
```
For all state bits, might be worth defining `static inline` (or better `__attribute__((always_inline))`) or a MACRO to read the different flags. Might avoid some stupid bugs of shifting with numbers all over the codebase

rzalamena replied:
Sure, I'll add static inline functions in the header to do this.

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+	msg.header.type = htons(ECHO_REQUEST);
+
+	/* Payload data. */
+	gettimeofday(&tv, NULL);
```
So, might be important... might not.

`gettimeofday()` is affected by discontinous jumps in the clock wall (e.g. resync of NTP, manual changes on `date`).

I am not sure what tis this time used for, but if it's used for internal callbacks you might want to have a look into `CLOCK_MONOTONIC`, not affected by jumps and significantly cheaper. Take a look into `CLOCK_MONOTONIC_COARSE` too, but this has more portability issues.

rzalamena replied:
It is not used for critical things, only for measuring ping latency/processing delay.

We can't use MONOTONIC clock here because each application has its "own" MONOTONIC clock. It gets a bit more complicated when the data plane is running on a different OS/machine than the FRR BFD too.

msune replied:
You mean it's shared across different processes, potentially in different machines?

rzalamena replied:
Yes.

Here is the FRR BFD code that answers:
https://github.com/opensourcerouting/frr/blob/bfd-distributed-v3/bfdd/dplane.c#L463

msune replied:
ACK

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+	msg.header.type = htons(ECHO_REPLY);
+
+	/* Payload data. */
+	gettimeofday(&tv, NULL);
```
idem

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+ * Misc functions.
+ */
+uint64_t
+hu64tonu64(uint64_t value)
```
I don't fully understand these two functions.

rzalamena replied:
Sorry, I just figure that there is a function to convert 64bit variable endianess:
https://man7.org/linux/man-pages/man3/endian.3.html

I'll fix this.

msune commented on libbfddp/bfddp_extra.c:
```diff
@@ -0,0 +1,858 @@
+	if (dm_one)
+		jitter = (uint32_t)(random() % 11);
+	else
+		jitter = (uint32_t)(random() % 26);
```
Minor:

```
jitter = (uint32_t)(random() % (dm_one? 11 : 26));
```

Also consider using macros for `11`, `26` to know what these "magic" numbers mean

rzalamena replied:
I'll fix this.

msune commented on libbfddp/bfddp_extra.h:
```diff
@@ -0,0 +1,497 @@
+	/** Interface index. */
+	uint32_t bs_ifindex;
+	/** Interface name. */
+	char bs_ifname[64];
```
might be worth defining `BFD_IF_MAX_SIZE` or s/t like that

rzalamena replied:
Makes sense, I'll fix it.

msune commented on libbfddp/bfddp_extra.h:
```diff
@@ -0,0 +1,497 @@
+ * \param bcp the control packet buffer.
+ */
+void
+bfddp_fill_control_packet(const struct bfd_session *bs,
```
There is a mixture of code putting the return type in the same line or with a breakline before the function name. Might be worth homogenizing

rzalamena replied:
Nice catch, you are correct... I forgot to run `clang-format`.

msune commented on libbfddp/bfddp_packet.h:
```diff
@@ -77,7 +112,7 @@ struct bfddp_echo {
 enum bfddp_session_flag {
-	/** Set when using multihop. */
+	/** Set when using multi hop. */
 	SESSION_MULTIHOP = (1 << 0),
```
`BFDDP_SESSION_MULTIHOP`?

msune commented on libbfddp/bfddp_packet.h:
```diff
@@ -223,14 +257,11 @@ struct bfddp_state_change {
 enum bfddp_control_state_bits {
-	/** Used to request connection establishement signal. */
+	/** Used to request connection establishment signal. */
 	STATE_POLL_BIT = (1 << 5),
```
idem

msune commented on soft_bfddpd/bfddpd.c:
```diff
@@ -71,7 +64,8 @@ static void bfddp_handle_message(struct events_ctx *ec, struct bfddp_ctx *bctx);
 /**
  * Handle common read/write events.
  */
-static int bfddp_event(struct events_ctx *ec, int fd, short revents, void *arg);
```
would it be possible to move the impl. of these functions to here, to avoid unnecessary fwd decl.?

rzalamena replied:
I'd agree with this, but the code is already there and it would generate bigger changes (and possible conflicts) :(

msune commented on soft_bfddpd/events.c:
```diff
@@ -48,6 +47,15 @@ struct pollfd_ctx {
 
+enum timer_state_flag {
+	/** Mark as removed from timers tree. */
+	TSF_REMOVED = (1 << 0),
```
Prefix/

rzalamena replied:
Sorry I did not understand this comment.

msune replied:
Also prefix with `SBFD_` or s/t alike

rzalamena replied:
It is already prefixed with `TSF_` and it is not exported (only used internally)... I would recommend people to use battle tested event libraries instead of this (e.g. `libevent`, `libuv` etc...). The motivation was mostly to avoid external dependencies.

I even made some effort to avoid external data structures libraries, but I couldn't escape some.

msune commented:
@rzalamena Had some high level comments on generic things. I cannot judge the code itself. @peterd-sky might give more hints.

Good job

rzalamena commented:
> @rzalamena Had some high level comments on generic things. I cannot judge the code itself. @peterd-sky might give more hints.
> 
> Good job

Thank you for your feedback, I'll fix the issues and push to this branch later.

peterd-sky commented on libbfddp/bfddp_extra.h:
```diff
@@ -0,0 +1,497 @@
+ * BFD state change: it is called every time the session state changes.
+ *
+ * \param bs the BFD session.
+ * \param ostate the old BFD session state.
```
Missing Doxygen comment for `arg`

peterd-sky commented on libbfddp/bfddp_extra.h:
```diff
@@ -0,0 +1,497 @@
+ *
+ * \param bds the BFD message.
+ * \param arg application argument.
+ * \param bctx the FRR BFD data plane to talk to.
```
Doxygen comments for the parameters don't match the actual parameters.

peterd-sky commented on libbfddp/bfddp_extra.h:
```diff
@@ -0,0 +1,497 @@
+ * FRR's BFD daemon.
+ *
+ * \param bs the BFD session.
+ */
```
Missing Doxygen comment for `arg`

peterd-sky commented on libbfddp/bfddp_packet.h:
```diff
@@ -33,6 +33,41 @@
+
+/** BFD single hop UDP port, as defined in RFC 5881 Section 4. Encapsulation. */
+#define BFD_SINGLE_HOP_PORT 3784
+
```
Consider adding `#define BFD_SINGLE_HOP_ECHO_PORT 3785` or something like that just to be complete.

peterd-sky commented on soft_bfddpd/bfddpd.c:
```diff
@@ -139,30 +133,33 @@ parse_address(const char *arg, struct sockaddr *sa, socklen_t *salen)
 	sptr++;
-	memcpy(type, arg, typelen);
-	type[typelen] = 0;
+	memcpy(type, arg, slen);
```
Consider guarding against overwriting the boundary of type should slen be greater than 64.

rzalamena committed:
libbfddp: more control packet validations

On packet session processing include some more validations and return a
value to tell if the operation was successful or failure so the caller
may do error counting.

rzalamena committed:
libbfddp: readability improvement

Move macro outside of the function and remove `undef` as it is not
necessary in this context.

This commit has no functional change intended.

rzalamena committed:
libbfddp: assert on invalid states

Implement a library assertion and use it to validate state machine and
callbacks conditionals.

rzalamena committed:
libbfddp: code readability improvement

Quit function execution as soon as possible. No functional change
intended.

rzalamena committed:
libbfddp: simplify code on control packet receive

Don't use conditionals/branching when we can get away by just doing
assignments.

rzalamena committed:
libbfddp: wrap shift operations with functions

Lets avoid shifting bugs for us and our users by wrapping those into a
inline function.

rzalamena committed:
libbfddp: define jitter percentages

Don't use hard coded numbers, instead define them (explain where they
were took from) and use them in the context.

rzalamena committed:
libbfddp: replace yet another hard code number

Use a macro to define the theoretical maximum interface name size.

rzalamena committed:
libbfddp: fix some header issues

* Fix some missing documentation for function arguments.
* Run clang-format to fix prototype declaration style.

rzalamena committed:
libbfddp: replace endian conversion functions

Use the 64 bit conversion functions provided by `endian.h` instead of
hand rolling my own.

rzalamena committed:
soft_bfddpd: check for invalid long socket types

Socket types are short words anything else is bogus.

rzalamena commented:
@msune and @peterd-sky thank you for the reviews, I fixed the items you pointed out.

Please let me know if you want to point out more items or I'll merge this.

rzalamena committed:
libbfddp: add BFD echo port definition

It is not currently being used in libbfddp but lets put it there for
completeness.

rzalamena committed:
libbfddp: allow application to set socket

Allow application to have freedom to provide whatever it wants as file
descriptor for data plane context.
